runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java

```diff
@@ -38,36 +53,117 @@
 public class FlinkExecutableStageFunction<InputT>
     extends RichMapPartitionFunction<WindowedValue<InputT>, RawUnionValue> {
 
+  // Main constructor fields. All must be Serializable because Flink distributes Functions to
+  // task managers via java serialization.
+
   // The executable stage this function will run.
   private final RunnerApi.ExecutableStagePayload stagePayload;
   // Pipeline options. Used for provisioning api.
-  private final Struct pipelineOptions;
+  private final JobInfo jobInfo;
   // Map from PCollection id to the union tag used to represent this PCollection in the output.
   private final Map<String, Integer> outputMap;
+  private final SerializableSupplier<FlinkBundleFactory> bundleFactorySupplier;
+  private final DistributedCachePool.Factory cachePoolFactory;
+  private final FlinkStateRequestHandlerFactory stateHandlerFactory;
+
+  // Worker-local fields. These should only be constructed and consumed on Flink TaskManagers.
+  private transient RuntimeContext runtimeContext;
+  private transient StateRequestHandler stateRequestHandler;
+  private transient StageBundleFactory stageBundleFactory;
+  private transient AutoCloseable distributedCacheCloser;
 
   public FlinkExecutableStageFunction(
       RunnerApi.ExecutableStagePayload stagePayload,
-      Struct pipelineOptions,
-      Map<String, Integer> outputMap) {
+      JobInfo jobInfo,
+      Map<String, Integer> outputMap,
+      SerializableSupplier<FlinkBundleFactory> bundleFactorySupplier,
+      DistributedCachePool.Factory cachePoolFactory,
+      FlinkStateRequestHandlerFactory stateHandlerFactory) {
     this.stagePayload = stagePayload;
-    this.pipelineOptions = pipelineOptions;
+    this.jobInfo = jobInfo;
     this.outputMap = outputMap;
+    this.bundleFactorySupplier = bundleFactorySupplier;
+    this.cachePoolFactory = cachePoolFactory;
+    this.stateHandlerFactory = stateHandlerFactory;
   }
 
   @Override
   public void open(Configuration parameters) throws Exception {
-    throw new UnsupportedOperationException();
+    ExecutableStage executableStage = ExecutableStage.fromPayload(stagePayload);
+    runtimeContext = getRuntimeContext();
+    // NOTE: It's safe to reuse the state handler between partitions because each partition uses the
+    // same backing runtime context and broadcast variables. We use checkState below to catch errors
+    // in backward-incompatible Flink changes.
+    stateRequestHandler = stateHandlerFactory.forStage(executableStage, runtimeContext);
+    DistributedCachePool cachePool = cachePoolFactory.forJob(jobInfo.jobId());
+    distributedCacheCloser = cachePool.addCacheToPool(runtimeContext.getDistributedCache());
```

Comment:
* Weren't we going to have a pool of ArtifactSource rather than a pool of DistributedCache?
* This pooling business is very funky (remember how much in-person discussion it took to remember why we even need it), could you add some comments about what's going on here?

Reply:
Done. I've added those comments in ArtifactSourcePool.